ring.go

```diff
@@ -213,32 +215,71 @@ type ringShards struct {
 
 	mu       sync.RWMutex
 	hash     ConsistentHash
-	shards   map[string]*ringShard // read only
-	list     []*ringShard          // read only
+	shards   map[string]*ringShard // read only, updated by SetAddrs
+	list     []*ringShard          // read only, updated by SetAddrs
 	numShard int
 	closed   bool
 }
 
 func newRingShards(opt *RingOptions) *ringShards {
-	shards := make(map[string]*ringShard, len(opt.Addrs))
-	list := make([]*ringShard, 0, len(shards))
+	c := &ringShards{
+		opt: opt,
+	}
+	c.SetAddrs(opt.Addrs)
 
-	for name, addr := range opt.Addrs {
-		shard := newRingShard(opt, name, addr)
-		shards[name] = shard
+	return c
+}
 
-		list = append(list, shard)
+// SetAddrs must not be called concurrently.
+func (c *ringShards) SetAddrs(addrs map[string]string) {
+	shards := make(map[string]*ringShard)
+	if c.closed {
+		return
+	}
+
+	for k, shard := range c.shards {
+		if addr, ok := addrs[k]; ok && shard.addr == addr {
+			shards[k] = shard
+		} else {
+			// close shards that are not reused
+			defer func() {
+				c.mu.Lock()
+				if !c.closed {
+					err := shard.Client.Close()
+					if err != nil {
+						internal.Logger.Printf(context.Background(), "Failed to close ring shard client %s %s: %v", k, shard.addr, err)
+					}
+				}
+				c.mu.Unlock()
+			}()
```

Comment:
I think it would be cleaner to track `unusedShards` and close them at the end - this would be consistent with closing `newShardNames`.